src/OpenTelemetry.Contrib.Instrumentation.Remoting/Implementation/TelemetryDynamicSink.cs

```diff
@@ -0,0 +1,127 @@
+
+        public void ProcessMessageStart(IMessage reqMsg, bool bCliSide, bool bAsync)
+        {
+            // Are we executing on client?
+            if (bCliSide)
+            {
+                // Start new outgoing activity
+                var act = RemotingActivitySource.StartActivity(ActivityOutName, ActivityKind.Client);
+                if (act != null)
+                {
+                    act.SetTag("uri", reqMsg.Properties["__Uri"]);
+
+                    // TODO: other useful tags (method name? parameters?)
+
+                    var callContext = (LogicalCallContext)reqMsg.Properties["__CallContext"];
+
+                    Propagator.Inject(new PropagationContext(act.Context, Baggage.Current), callContext, InjectActivityProperties);
+                }
+            }
+            else
+            {
+                // We are on server, need to start new or attach to an existing incoming activity.
+
+                // Do we already have an incoming activity?
+                // Existing activity might be started by an instrumentation layer higher up the
+                // stack. E.g. if we are using http as a transport for remoting, and we have
+                // instrumented incoming http request, then that instrumentation happens before
+                // our "ProcessMessageStart" and we just need to attach to that existing activity.
+                var act = Activity.Current;
+                if (act != null)
+                {
+                    RemotingActivitySource.StartActivity(ActivityInName, ActivityKind.Server);
+
+                    // TODO: "merge" things that have been passed via __CallContext, add useful remoting tags
+                }
+                else
+                {
+                    // We don't have an existing incoming activity. Start a brand new one ourselves.
+
+                    var callContext = (LogicalCallContext)reqMsg.Properties["__CallContext"];
+
+                    var activityParentContext = Propagator.Extract(default, callContext, ExtractActivityProperties);
+
+                    RemotingActivitySource.StartActivity(ActivityInName, ActivityKind.Server, activityParentContext.ActivityContext);
+                }
+            }
+        }
+
+        public void ProcessMessageFinish(IMessage replyMsg, bool bCliSide, bool bAsync)
+        {
+            // This will be called:
+            // 1) On the server after we are done processing the call and are returning back to client
+            //    => Current activity should be the "ActivityIn", we need to stop it
+            //
+            // 2) On the client, when the server call returns
+            //    => Current activity should be the "ActivityOut", we need to stop it as well
+
+            // TODO: is this wrong? Don't just stop _any_ current activity, check if it is either of the above explicitly?
+            var act = Activity.Current;
```

Comment:
In general, I think the idea is: "Stop what you started, and nothing else."

Reply:
Thanks, makes sense! I've added checks based on operation type and `activity.DisplayName` similar to how it was done in ASP.NET instrumentation I think.

It is still possible to have something like that:
* Instrumentation starts `Remoting-In` activity
* Server code starts child `aServerActivity` and doesn't stop it
* Remoting call completes, `ProcessMessageFinish` fires, current activity is still `aServerActivity`...

It is probably possible to capture the `Remoting-In` activity in the call-context and then stop it on `ProcessMessageFinish`, but the still-active `aServerActivity` is likely an error (missing Stop/Dispose) in the server code, so perhaps it is ok to log it and move on?

Reply:
> Server code starts child aServerActivity and doesn't stop it

Do you means Asp.Net starting its activity, and not stopping it?